core/src/main/java/org/apache/calcite/rel/rules/JoinConditionOrExpansionRule.java

```diff
@@ -192,6 +185,27 @@ private boolean isJoinKeyCond(RexCall call, int leftFieldCount) {
     return false;
   }
 
+  private boolean containNonOrOneInputRef(RexNode rex) {
+    RexInputRefCounter counter = new RexInputRefCounter();
+    rex.accept(counter);
+    return counter.inputRefCount < 2;
+  }
+
+  /**
+   * Counts the number of InputRefs in a RexNode expression. */
+  private static class RexInputRefCounter extends RexVisitorImpl<Void> {
+    public int inputRefCount = 0;
```

Comment:
counting RexInputRef is not what you want.
What you want to see is whether the reference is in the left or right input of the join.
So you need to pass the leftFieldCount.

Reply:
I think an expression like `left.a = left.b` is still "legal", and should be extracted, because after extraction it can be pushed up to the left input, if I understand how this expansion will work